### PR TITLE
Refactor/economy

### PR DESCRIPTION
**Pull request by suneettipirneni**
#### Description: 

# Motivation
The motivation may seem a bit elusive at first, as detailed explanation is required. We should kick off this explanation with a review of the *form* of objects rather than the traditional perspective of *function*.

### Let's see an example.
I'm going to use the route of inheritance via objects to demonstrate an issue.
Observe the following code:
```java
public class Animal {
    ...
}
public class AnimalDiet extends Animal {
    ...
}
// A Panda has a Diet
public class Panda extends AnimalDiet {
    ...
}
// A Bear has a Diet
public class Bear extends AnimalDiet {
    
}
```
So for the code above fits our needs, now let's introduce a catalyst for failure:
```java
public class CarnivorousAnimal extends Animal {
    ...
}
``` 
But now I need my `Bear` class to extend `CarnivorousAnimal` because it is also a carnivore *and* has a diet. Now I'm going to have to use an inheritance chain:

```java
public class CarnivorousAnimal extends AnimalDiet {
    ...
}

public class Bear extends CarnivorousAnimal {
    ...
}
```
We now have the problem where we've locked `CarnivorousAnimal` into objects that *only* have *both* a diet and are carnivores. This is where this kind of design breaks down, it's not future proof, and it's highly static as you need to change superclasses in order to make changes in subclasses. So what's the solution to this?

There's actually two:
1. Create an interface for each attribute (Good if there is variation between classes)
2. Instead of using class inheritance make the attributes of the class in the form of a field. (Good if there isn't much variation in implementation)

So let's fix the code, (I'm going to use a mix of method 1 and 2 for this):
```java
public class Animal {
    ...
}
public class Diet {
    ...
}
public interface Dietable {
    Diet getDiet();
}
public interface Carnivorous {
    ...
}
public class Bear extends Animal implements Carnivorous, Dietable {
    private Diet diet;
    
    @Override 
    public Diet getDiet() {
        return diet;
    }
}
```
Notice how now the clarity of the object inheritance has been restored whilst at the same time allowing for future changes to be easier.

This same idea has been applied to `TownyEconomyObject`, for the same or similar reasons as outlined above. In addition, new interfaces have been introduced to further *declutter* a rather messy inheritance chain.

New Interfaces:
- `TownBlockOwner` - Now applied to Resident and Town to generify uses.
- `EconomyHandler` - Used to show that an object can handle economic interactions.
- `Permissible` - Used to show that an object can handle permission interactions.

With the addition of these interfaces the main Towny objects now directly inherit from `TownyObject` providing a clear and concise object hierarchy that is not only aesthetically pleasing but also modular enough for future expansion.
____
- [x] I have tested this pull request for defects on a server. (**Reccomended that others test as well**)

By making this pull request, I represent that I have the right to waive copyright and related rights to my contribution, and agree that all copyright and related rights in my contributions are waived, and I acknowledge that the TownyAdvanced organization has the copyright to use and modify my contribution under the Towny [License](https://github.com/LlmDl/Towny/blob/master/LICENSE.md) for perpetuity.

### CODE REVIEW

File: src/com/palmergames/bukkit/towny/object/Nation.java [hunk 1]
```diff
@@ -565,21 +566,6 @@ public boolean hasResident(String name) {
...
-            return super.getBukkitWorld();
-        }
-    }
-
-
-	@Override
-	public String getEconomyName() {
-		return StringMgmt.trimMaxLength(Nation.ECONOMY_ACCOUNT_PREFIX + getName(), 32);
-	}
-
```
Comment (LlmDl) [inline src/com/palmergames/bukkit/towny/object/Nation.java]: Why are these getting nuked? There will certainly be plugins using the getEconomyName() to manipulate the eco account.

File: src/com/palmergames/bukkit/towny/object/EconomyAccount.java [hunk 2]
```diff
@@ -1,201 +1,220 @@
...
-
-	/**
-	 * Attempt to delete the economy account.
-	 */
-	public void removeAccount() {
-		TownyEconomyHandler.removeAccount(getEconomyName());
-	}
-
-}
+package com.palmergames.bukkit.towny.object;
```
Comment (LlmDl) [inline src/com/palmergames/bukkit/towny/object/EconomyAccount.java]: This file got formatted throughout.

File: src/com/palmergames/bukkit/towny/object/Town.java [hunk 3]
```diff
@@ -1051,21 +1079,7 @@ public boolean isPublic() {
...
-        } else {
-            return super.getBukkitWorld();
-        }
-    }
-
-	@Override
-	public String getEconomyName() {
-		return StringMgmt.trimMaxLength(Town.ECONOMY_ACCOUNT_PREFIX + getName(), 32);
-	}
-
```
Comment (LlmDl) [inline src/com/palmergames/bukkit/towny/object/Town.java]: Same as nation, these should probably be left in, maybe deprecated.

**Comment by suneettipirneni** [comment]
Should be good now

**Pull request merged**